python/ray/train/base_trainer.py

```diff
@@ -351,25 +413,41 @@ def fit(self) -> Result:
             TrainingFailedError: If any failures during the execution of
             ``self.as_trainable()``.
         """
-        from ray.tune.tuner import Tuner
-        from ray.tune.error import TuneError
+        from ray.tune.tuner import Tuner, TunerInternal
 
         trainable = self.as_trainable()
         param_space = self._extract_fields_for_tuner_param_space()
 
-        tuner = Tuner(
-            trainable=trainable, param_space=param_space, run_config=self.run_config
-        )
+        if self._restore_path:
+            tuner = Tuner.restore(
+                self._restore_path,
+                overwrite_trainable=trainable,
+                resume_unfinished=True,
+                resume_errored=True,
+            )
+        else:
+            tuner = Tuner(
+                trainable=trainable, param_space=param_space, run_config=self.run_config
+            )
+            experiment_path = Path(
+                TunerInternal.setup_create_experiment_checkpoint_dir(
+                    trainable, self.run_config
+                )
+            )
+            self._save(experiment_path)
+
         result_grid = tuner.fit()
         assert len(result_grid) == 1
-        try:
-            result = result_grid[0]
-            if result.error:
-                raise result.error
-        except TuneError as e:
-            raise TrainingFailedError from e
+        result = result_grid[0]
+        if result.error:
+            raise TrainingFailedError from result.error
```

Comment:
I changed this because the old try/catch did not most errors raised by the trial, and I was asserting for `TrainingFailedError` in my tests.

Reply:
Reverted this. Will do this in a separate PR.